synapse/lib/blobfile.py

```diff
@@ -0,0 +1,310 @@
+
+            # Write the size dword
+            self.writeoff(self._size, struct.pack(headfmt, size))
+
+            # Compute the return value
+            dataoff = self._size + headsize
+
+            # Update our runtime size
+            self._size = nsize
+
+        return dataoff
+
+    def size(self):
+        '''
+        Get the amount of space currently used by the blob.
+
+        Returns:
+            int: Size of the bytes currently allocated by the blob.
+        '''
+        return self._size
+
+    def walk(self, callback=None):
+        '''
+        Walk the blobfile from the first record until the end.
+
+        Args:
+            callback (func): A callback function which is executed when a
+            BlobFile record is encountered.  The callback should accept the
+            following function signature: ``(fd, baseoff, off, size)``.  fd is
+            the blob's file descriptor; baseoff is the offset where the blob
+            header is located, offset is the location of the blob data, and
+            size is the number of bytes allocated for the blob. If no callback
+            is provided, nothing will happen when a BlobFile record is
+            encountered.
+
+        Notes:
+            In addition to the callback, when walking the blobfile additional
+            conditions may be encountered. These will fire an event and exit
+            the walk() operation.
+
+            These events will not serialize since they contain a
+            reference to the underlying file descriptor and are designed in
+            order to allow creation of tools which can inspect/consume bytes
+            from a BlobFile.
+
+                blob:walk:unpkerr - Fired if there is an error unpackting the Blobfile header.
+
+                blob:walk:truncated - Fired if there is a truncated file encountered.
+
+                blob:walk:done - Fired when the BlobWalker is done walking the file.
+
+        Returns:
+            None
+        '''
+        # Snag the max filesize
+        self.fd.seek(0, os.SEEK_END)
+        maxsize = self.fd.tell()
+
+        # Start walking at zero
+        self.fd.seek(0)
+        while True:
+
+            # Store the current blobs offset value
+            baseoff = self.fd.tell()
+
+            # Read the header
+            try:
+                header = self.fd.read(headsize)
+                size, = struct.unpack(headfmt, header)
+            except Exception as e:
+                log.exception('Failed to read/unpack header')
+                self.fire('blob:walk:unpkerr',
+                          fd=self.fd,
+                          baseoff=baseoff,
+                          size=headsize,
+                          excinfo=s_common.excinfo(e))
+                break
+
+            # Save the current fd position
+            off = self.fd.tell()
+
+            if callback:
+                callback(fd=self.fd, baseoff=baseoff, off=off, size=size)
```

Comment:
Rather than specifying a callback, lets make this a yield generator function that yields: (off, size) and assumes the consumer may call readoff() if desired ( which safely uses self.atom.readoff correctly )